### DISCUSSION TITLE
Troubles with a pesky (well understood) memory leak (protobuf/cpputest)

### DISCUSSION THREAD

**Discussion by swinchen** [cpputest/cpputest#1828] [Q&A]
Hi all,

CppUTest 4.0
Protobuf 3.18.3
GCC 11.2.0

I am writing tests for a RPC client that is using protobuf under the hood - here is a very simple protofile to demonstrate the issue I am experiencing:

```protobuf
syntax = "proto3";
import "google/protobuf/any.proto";

package example.memleak;

message Payload {
  uint32 dummy = 1;
}

message Container {
  repeated google.protobuf.Any inner = 1;
}
```

And here is the test:
```cpp

static void fill_container(example::memleak::Container *container)
{
    example::memleak::Payload payload;
    payload.set_dummy(100);

    google::protobuf::Any *new_inner = container->add_inner();
    new_inner->PackFrom(payload);    
}

TEST(test_group, test_memleak)
{
    example::memleak::Container container;
    fill_container(&container);
    container.clear_inner();
}
```

This causes a leak because internally protobuf allocates pools of memory that it doesn't free until the program terminates.
I can fix this error by calling `google::protobuf::ShutdownProtobufLibrary();` but after this called all other tests using protobuf will segfault.

I tried adding `IGNORE_ALL_LEAKS_IN_TEST();` but this seems to have no effect.

I have also tried wrapping the test with `MemoryLeakWarningPlugin::saveAndDisableNewDeleteOverloads();` and `MemoryLeakWarningPlugin::restoreNewDeleteOverloads();` this causes a segfault - I haven't tracked down exactly why, but I suspect it is a mismatch between the version of new and delete used seeing protobuf might be freeing memory after the New/Delete overloads have been restored.

I am not sure if anyone has and suggestions or ideas on what to try next, but any advice would be appreciated.
Thanks.

-- edit
Added version numbers

**Comment by basvodde** [comment]
Thanks. I replied via the google group. Sorry for delay.
